daffodil-runtime1/src/main/java/org/apache/daffodil/udf/UDFunctionService.java

```diff
@@ -0,0 +1,137 @@
+
+				List<Class<?>> missingAnnotations = fc.stream()
+						.filter(f -> !f.isAnnotationPresent(FunctionClassInfo.class)).collect(Collectors.toList());
+
+				List<Class<?>> nonSerializables = fc.stream().filter(f -> Serializable.class.isAssignableFrom(f))
+						.collect(Collectors.toList());
+
+				if (!missingAnnotations.isEmpty()) {
+					warnings.add(
+							String.format("Provider ignored: [%s].\nAnnotations missing for FunctionClass(es):\n%s",
+									fcp.getClass().getName(), missingAnnotations.stream().map(c -> c.getName())
+											.collect(Collectors.joining("\n", "\t[", "]"))));
+					return;
+				}
+
+				if (!nonSerializables.isEmpty()) {
+					warnings.add(String.format(
+							"Provider ignored: [%s].\nFunctionClass(es) must implement java.io.Serializable:\n%s",
+							fcp.getClass().getName(), nonSerializables.stream().map(c -> c.getName())
+									.collect(Collectors.joining("\n", "\t[", "]"))));
+					return;
+				}
+
+				fc.stream().forEach(f -> {
+					FunctionClassInfo fInfo = f.getAnnotation(FunctionClassInfo.class);
+					String ns = fInfo.namespace();
+					String fname = fInfo.name();
+					if (ns.isEmpty() || fname.isEmpty() || ns.matches("\\s+") || fname.matches("\\s+")) {
+						warnings.add(String.format("FunctionClass ignored: [%s]. From Provider [%s].%s%s", f.getName(), fcp.getClass().getName(),
+								ns.isEmpty() ? "\nAnnotation namespace field is empty or invalid." : "",
+								fname.isEmpty() ? "\nAnnotation name field is empty or invalid." : ""));
+					}
+
+					String key = String.join("_", fInfo.namespace(), fInfo.name());
+					if (functionProviderLookup.containsKey(key)) {
+						warnings.add(String.format(
+								"FunctionClass ignored: [%s].\nDuplicate name [%s] and namespace[%s] found.",
+								f.getName(), key.split("_")[0], key.split("_")[1]));
+					} else {
+						functionProviderLookup.put(key, fcp);
+						functionClasses.add(f);
+					}
+				});
+			});
+
+			if (functionClasses.isEmpty()) {
+				errors.add(String.format("No user defined functions found."
+						+ "\nCheck that UDF JARs are on classpath and that they are properly registerable by ServiceLoader."
+						+ "\nCurrent classpath locations:\n%s",
+						Arrays.asList(Misc.getClassPath()).stream().map(u -> u.toString())
+								.collect(Collectors.joining("\n"))));
+			}
+		} catch (ServiceConfigurationError | Exception e) {
+			e.printStackTrace();
+			errors.add(e.toString() + String.format("\nCurrent classpath locations:\n%s", Arrays
+					.asList(Misc.getClassPath()).stream().map(u -> u.toString()).collect(Collectors.joining("\n"))));
+		}
+	}
+
+	public Object lookupFunctionClass(String namespace, String name) {
+		Object funcObj = null;
+
+		try {
+			UDFunctionProvider fp = this.functionProviderLookup.get(String.join("_", namespace, name));
+			if (fp != null) {
+				funcObj = fp.lookupFunctionClass(namespace, name);
+			}
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+		return funcObj;
+	}
```

Comment:
If the same UDFunction is called multiple times we'll call lookupFunctionClass each time and presumably get a new instance. Should we cache the original instance and reuse it, or is it up to the caller to cahe the instance and give us the same one if it's important to them?

Reply:
Hmm, if we consider state, we'd probably want it to be up to the provider depending on whether they want state refreshed each time, or cached, so I think it should be up to them to return what they want to us based on need.

Reply:
My expectation as a user would be that Daffodil caches the functions. If a function has a significant initialization cost, the path of least resistance for the user would still be to rebuild it for every request. Since we do not support mutable state at all, I do not see what benifit letting the user rebuild the objects gives us.

If, at some point in the future, we decide to allow some form of mutable state, we would probably need to add some life-cycle methods to the function objects, which would preclude re-initalizing them at each call point as well.

Reply:
Ok, sounds good, I'll update